### PR TITLE
Test Thoughts & Subthoughts

### PR DESCRIPTION
**Pull request by mekanix**
Refers to #244

### CODE REVIEW

**Comment by mekanix** [comment]
Only first test is passing, no matter which test it is. I'm suspecting I need to clean up something or reinit firebase or something, so I'm closing this PR for now.

**Comment by mekanix** [comment]
Everything works if we have one `it` section per test file. As test file naming scheme doesn't have to follow component naming, we can have as many test files as we want, so I think it's an acceptable trade-off.

**Comment by raineorshine** [comment]
I am unfortunately getting the following error when I try to run the tests:

```
> em@85.0.0 test /Users/raine/projects/em
> react-scripts test --env=jsdom --silent

2020-02-03 11:54 node[66157] (FSEvents.framework) FSEventStreamStart: register_with_server: ERROR: f2d_register_rpc() => (null) (-22)
2020-02-03 11:54 node[66157] (FSEvents.framework) FSEventStreamStart: register_with_server: ERROR: f2d_register_rpc() => (null) (-22)
2020-02-03 11:54 node[66157] (FSEvents.framework) FSEventStreamStart: register_with_server: ERROR: f2d_register_rpc() => (null) (-22)
events.js:187
      throw er; // Unhandled 'error' event
      ^

Error: EMFILE: too many open files, watch
    at FSEvent.FSWatcher._handle.onchange (internal/fs/watchers.js:123:28)
Emitted 'error' event on FSWatcher instance at:
    at FSEvent.FSWatcher._handle.onchange (internal/fs/watchers.js:129:12) {
  errno: -24,
  syscall: 'watch',
  code: 'EMFILE',
  filename: null
}
npm ERR! Test failed.  See above for more details.
```

**Comment by raineorshine** [comment]
Looking into a solution, as it may be OSX specific.

https://github.com/facebook/create-react-app/issues/4540

**Comment by raineorshine** [comment]
Sweet, installing watchman was all it took to fix that.

File: package.json [hunk 1]
```diff
@@ -64,7 +65,7 @@
     "deploy": "npm version major && npm run build && firebase deploy",
     "lint": "eslint src",
     "pretest": "npm run lint",
-    "test": "react-scripts test",
+    "test": "react-scripts test --env=jsdom --silent",
```
Comment (raineorshine) [inline package.json]: `--silent` also prevents any console.logs in tests from outputting, which can make debugging hard. I would suggest removing this and either living with the warnings or finding a way to suppress them.

**Comment by mekanix** [comment]
> Sweet, installing watchman was all it took to fix that.

Do you want me to add it as dev dependency?

File: src/components/__tests__/EmptyThought.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import React from 'react'
+import { App } from '../App'
+
+// eslint-disable-next-line fp/no-let
+let wrapper = null
+
+// eslint-disable-next-line no-undef
+beforeAll(async () => {
```
[the remark this answers is not in the document]
Comment (mekanix) [reply src/components/__tests__/EmptyThought.js]: But that will add "jest" as env even in builds, if I'm not wrong. Do we really want that?

File: src/util/initEvents.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ export const initEvents = () => {
   // prevent browser from restoring the scroll position so that we can do it manually
   window.history.scrollRestoration = 'manual'
 
-  window.addEventListener('keydown', handleKeyboard)
```
[the remark this answers is not in the document]
Comment (mekanix) [reply src/util/initEvents.js]: Handlers have to be bound to some element/component (not window or document) or tests have no way of using them. I'll look into Search and New Thought soon.

File: src/components/__tests__/EmptyThought.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (raineorshine) [reply src/components/__tests__/EmptyThought.js]: Are there other implications I am not aware of? It seems to work fine for me.

**Comment by raineorshine** [comment]
> Do you want me to add it as dev dependency?

No, too heavy. Developers can install it if needed for now.

File: src/components/__tests__/EmptyThought.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mekanix) [reply src/components/__tests__/EmptyThought.js]: I didn't search for what it means for the browser/mobile, so maybe it's fine. I was just thinking less code has less chance of breaking something.

File: src/util/initEvents.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (raineorshine) [reply src/util/initEvents.js]: It might be feasible to attach `handleKeyboard` to a component for testing purposes while attaching it to `window` during normal execution of the app. 

In the tests, we can always dispatch the action directly (e.g. `{ type: 'search', value: 'TEXT' }`) rather than simulating the event. 

Note that we should *not* do anything specific for Search and New Thought, but rather have a general solution.

**Comment by mekanix** [comment]
I left `--env=jest` in the package.json as without it there is no `window` or `document` objects, although there's a section in the `.eslintrc.json` that clearly states:
```
  "env": {
    "browser": true,
    "es6": true,
    "jest": true,
    "mocha": true
  },
```

File: src/components/App.js [hunk 4]
```diff
@@ -23,13 +22,6 @@ const HTML5toTouch = {
...
 export const App = DragDropContext(MultiBackend(HTML5toTouch))(() =>
   <Provider store={store}>
-    <div onTouchMove={
-      () => globals.touching = true // eslint-disable-line no-return-assign
-    } onTouchEnd={() => {
-      globals.touching = false // eslint-disable-line no-return-assign
-      globals.touched = true // eslint-disable-line no-return-assign
-    }}>
-      <AppComponent/>
-    </div>
```
Comment (raineorshine) [inline src/components/App.js]: If you remove this from the app and only add it in the tests, then it will not work in the build.

Comment (raineorshine) [reply src/components/App.js]: FWIW, I think these global touch variables can be safely ignored by the tests.

File: src/components/__tests__/EmptyThought.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  wrapper = await window.createTestApp(<App />)
+})
+
+afterAll(() => {
+  window.cleanupTestApp()
+  wrapper = null
+})
+
+it('create and edit empty thought', async () => {
```
Comment (raineorshine) [inline src/components/__tests__/EmptyThought.js]: Create and edit should be separate tests.

File: src/components/__tests__/EmptyThought.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import React from 'react'
+import App from '../../App'
+
+// eslint-disable-next-line fp/no-let
+let wrapper = null
+
+beforeAll(async () => {
+  wrapper = await window.createTestApp(<App />)
```
Comment (raineorshine) [inline src/components/__tests__/EmptyThought.js]: Can `createTestApp` and `cleanupTestapp` be imported directly, or do they have to be attached to `window`?

File: src/components/__tests__/EmptyThought.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+afterAll(() => {
+  window.cleanupTestApp()
+  wrapper = null
+})
+
+it('create and edit empty thought', async () => {
+  const thoughts = wrapper.find('div.transformContain div ul.children')
+  const thought = thoughts.find(
+    'li.leaf div.thought-container div.thought div.editable',
```
Comment (raineorshine) [inline src/components/__tests__/EmptyThought.js]: Omit tag names, i.e. `.leaf .thought-container .thought .editable`. In this case I think we can simply select `.editable`.

File: src/components/__tests__/EmptyThought.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  )
+  expect(thought.text()).toBe('')
+  const keyboardResponder = wrapper.find('div#keyboard')
```
Comment (raineorshine) [inline src/components/__tests__/EmptyThought.js]: `#keyboard`

File: src/components/__tests__/SubThought.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  wrapper = await window.createTestApp(<App />)
+})
+
+afterAll(() => {
+  window.cleanupTestApp()
+  wrapper = null
+})
+
+it('create and edit subthought', async () => {
+  const thoughts = wrapper.find('div.transformContain div ul')
```
Comment (raineorshine) [inline src/components/__tests__/SubThought.js]: `.transformContain` is not relevant for this test. We should try to write selectors that are semantically relevant for the current test.

File: src/components/__tests__/SubThought.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  await thought.update()
+  expect(thought.text()).toBe('c')
+  // eslint-disable-next-line fp/no-let
+  let subthoughtsData = getThoughtsRanked(['c'])
+  // eslint-disable-next-line fp/no-let
+  let subthoughtData = subthoughtsData[0]
+  expect(subthoughtData.value).toBe('')
+  await wrapper.update()
+  const subthought = wrapper.find(
+    'ul.distance-from-cursor-0 li.leaf div.thought-container div.thought div.editable',
```
Comment (raineorshine) [inline src/components/__tests__/SubThought.js]: In general, we want the minimal, semantic selector that identifies the relevant element for the test. This will be more resilient to markup changes.

File: src/components/App.js (hunk 4 above)
Comment (mekanix) [reply src/components/App.js]: I deleted it by mistake, it's back now

File: src/components/__tests__/EmptyThought.js (hunk 6 above)
Comment (mekanix) [reply src/components/__tests__/EmptyThought.js]: Done

File: src/components/__tests__/EmptyThought.js (hunk 7 above)
Comment (mekanix) [reply src/components/__tests__/EmptyThought.js]: I tried it, but tests fail that way

File: src/components/__tests__/EmptyThought.js (hunk 5 above)
Comment (mekanix) [reply src/components/__tests__/EmptyThought.js]: Think about it, it's useless. In order to edit the thought, you need to create it, so why not test it? Same for delete.

File: src/components/__tests__/SubThought.js (hunk 10 above)
Comment (mekanix) [reply src/components/__tests__/SubThought.js]: I tried. I even tried selectors with components. If you do find selector that is more general and works, please do tell!

File: src/components/__tests__/EmptyThought.js (hunk 5 above)
Comment (raineorshine) [reply src/components/__tests__/EmptyThought.js]: I disagree. If there is one test for creating a thought and for editing a thought, and the test fails after some changes are made in the future, we won't know whether there was a regression in creating a thought or editing a thought. While editing a thought will never work if creating a thought is broken, the inverse is not true. Having separate tests isolates these test cases and prevents this ambiguity from arising.

File: src/components/__tests__/SubThought.js (hunk 9 above)
Comment (mekanix) [reply src/components/__tests__/SubThought.js]: Same here, I tried.

File: src/components/__tests__/SubThought.js (hunk 10 above)
Comment (raineorshine) [reply src/components/__tests__/SubThought.js]: Okay, thanks for trying!

File: src/components/__tests__/EmptyThought.js (hunk 5 above)
Comment (mekanix) [reply src/components/__tests__/EmptyThought.js]: But failing test tells you exact line where it broke, so we will know what exactly failed. If you still want me to have create and edit, no problem.

Comment (raineorshine) [reply src/components/__tests__/EmptyThought.js]: Good point. Either one is a reasonable approach. Let's go with separate tests since that is my preference.

File: src/setupTests.js [hunk 11]
```diff
@@ -1,4 +1,77 @@
...
+        tabIndex="0"
+      >
+        {app}
+      </div>,
+      { attachTo: document.body },
+    )
+    const skipTutorial = wrapper.find('div.modal-actions div a')
+    skipTutorial.simulate('click')
+    const keyboardResponder = wrapper.find('#keyboard')
+    await keyboardResponder.simulate('keydown', { key: 'Enter' })
```
Comment (raineorshine) [inline src/setupTests.js]: Can we put the Enter event in the tests? I'm okay with the skipTutorial click happening here, but the new thought should go in each test. If a regression is introduced to creating a new thought, we want the newThought test to fail, not setupTests.

File: src/setupTests.js [hunk 12]
```diff
@@ -1,4 +1,77 @@
...
+    removeAllRanges: () => {},
+    addRange: () => {},
+  })
+
+  window.location = {
+    pathname: '',
+  }
+
+  await act(async () => {
+    // eslint-disable-next-line no-undef
```
Comment (raineorshine) [inline src/setupTests.js]: I believe you can remove the eslint-disable lines related to jest now that we have `jest` specified as an eslint env.

File: src/components/__tests__/ThoughtCreate.js [hunk 13]
```diff
@@ -0,0 +1,6 @@
+it('create thought', async () => {
+  const thought = document.wrapper.find(
+    'li.leaf div.thought-container div.thought div.editable',
```
Comment (raineorshine) [inline src/components/__tests__/ThoughtCreate.js]: Based on my tests, `document.wrapper.find('div.editable')` works fine.

`.editable` does not work because there is both a `<ContentEditable className='editable'>` and a `<div className='editable'>`

File: src/components/__tests__/ThoughtDelete.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import { act } from 'react-dom/test-utils'
+
+it('delete thought', async () => {
```
Comment (raineorshine) [inline src/components/__tests__/ThoughtDelete.js]: This looks good, and we will need a test for the `delete` shortcut which deletes a non-empty thought.

File: src/components/__tests__/SubThoughtCreate.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  )
+  expect(thought.text()).toBe('')
+  const keyboardResponder = document.wrapper.find('div#keyboard')
+  await thought.simulate('change', { target: { value: 'c' } })
+  await keyboardResponder.simulate('keydown', { key: 'Enter', ctrlKey: true })
+  jest.runAllTimers()
+  await thought.update()
+  expect(thought.text()).toBe('c')
+  // eslint-disable-next-line fp/no-let
+  const subthoughtsData = getThoughtsRanked(['c'])
```
Comment (raineorshine) [inline src/components/__tests__/SubThoughtCreate.js]: These variables are short enough that they can be inlined in the `expect`.

File: src/components/__tests__/SubThoughtDelete.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  )
+  expect(thought.text()).toBe('')
+  const keyboardResponder = document.wrapper.find('div#keyboard')
+  await thought.simulate('change', { target: { value: 'c' } })
+  await keyboardResponder.simulate('keydown', { key: 'Enter', ctrlKey: true })
+  jest.runAllTimers()
+  await thought.update()
+  expect(thought.text()).toBe('c')
+  // eslint-disable-next-line fp/no-let
+  let subthoughtsData = getThoughtsRanked(['c'])
```
Comment (raineorshine) [inline src/components/__tests__/SubThoughtDelete.js]: Let's avoid `let` here, and simply inline these short statements as needed.

File: src/components/__tests__/SubThoughtEdit.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    'li.leaf div.thought-container div.thought div.editable',
+  )
+  expect(thought.text()).toBe('')
+  const keyboardResponder = document.wrapper.find('div#keyboard')
+  await thought.simulate('change', { target: { value: 'c' } })
+  await keyboardResponder.simulate('keydown', { key: 'Enter', ctrlKey: true })
+  jest.runAllTimers()
+  await thought.update()
+  expect(thought.text()).toBe('c')
+  // eslint-disable-next-line fp/no-let
```
Comment (raineorshine) [inline src/components/__tests__/SubThoughtEdit.js]: Similar

File: src/components/__tests__/SubThoughtTop.js [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    'li.leaf div.thought-container div.thought div.editable',
+  )
+  expect(thought.text()).toBe('')
+  const keyboardResponder = document.wrapper.find('div#keyboard')
+  await thought.simulate('change', { target: { value: 'c' } })
+  await keyboardResponder.simulate('keydown', { key: 'Enter', ctrlKey: true })
+  jest.runAllTimers()
+  await thought.update()
+  expect(thought.text()).toBe('c')
+  // eslint-disable-next-line fp/no-let
```
Comment (raineorshine) [inline src/components/__tests__/SubThoughtTop.js]: Also here

**Comment by raineorshine** [comment]
Also let's move the `__tests__` folder up a level since technically we are testing more than components.

**Comment by mekanix** [comment]
I tried moving __tests__ one directory up, but this is what I end up with:
```
> em@86.0.0 lint /usr/src/em
> eslint src
​
​
> em@86.0.0 test /usr/src/em
> react-scripts test --env=jsdom
​
/usr/src/em/node_modules/react-scripts/scripts/test.js:20
  throw err;
  ^
​
Error: The `document` global was defined when React was initialized, but is not defined anymore. This can happen in a test environment if a component schedules an update from an asynchronous callback, but the test has already finished running. To solve this, you can either unmount the component at the end of your test (and ensure that any asynchronous operations get canceled in `componentWillUnmount`), or you can change the test itself to be asynchronous.
    at Object.invokeGuardedCallbackDev (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:298:17)
    at invokeGuardedCallback (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:440:31)
    at commitRootImpl (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:25020:9)
    at unstable_runWithPriority (/usr/src/em/node_modules/scheduler/cjs/scheduler.development.js:697:12)
    at runWithPriority$2 (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:12149:10)
    at commitRoot (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:24922:3)
    at finishSyncRender (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:24329:3)
    at performSyncWorkOnRoot (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:24307:9)
    at /usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:12199:24
    at unstable_runWithPriority (/usr/src/em/node_modules/scheduler/cjs/scheduler.development.js:697:12)
    at runWithPriority$2 (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:12149:10)
    at flushSyncCallbackQueueImpl (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:12194:7)
    at flushSyncCallbackQueue (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:12182:3)
    at batchedUpdates$1 (/usr/src/em/node_modules/react-dom/cjs/react-dom.development.js:24392:7)
    at Object.notify (/usr/src/em/node_modules/react-redux/lib/utils/Subscription.js:29:7)
    at Subscription.notifyNestedSubs (/usr/src/em/node_modules/react-redux/lib/utils/Subscription.js:71:20)
    at Subscription.handleChangeWrapper (/usr/src/em/node_modules/react-redux/lib/utils/Subscription.js:76:12)
    at dispatch (/usr/src/em/node_modules/redux/lib/redux.js:228:7)
    at Object.dispatch (/usr/src/em/node_modules/redux-thunk/lib/index.js:14:16)
    at /usr/src/em/src/util/loadLocalState.js:111:18
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
npm ERR! Test failed.  See above for more details.
```

**Comment by raineorshine** [comment]
> I tried moving tests one directory up, but this is what I end up with

Okay, that's too bad.

**Comment by raineorshine** [comment]
Thank you, those are perfectly sized and named commits.

**Comment by raineorshine** [comment]
Well, after a couple hours of trying different things myself, I was also unable to get the Editable element to update in the unit test. None of the tips in https://github.com/airbnb/enzyme/issues/76 seemed to work.

That said, the `change` event works great and the `store` is updated, so we can call `getThoughts` which accesses the store directly to see the changed value. Will you utilize this approach in all unit tests?

```js
import { ROOT_TOKEN } from '../../constants.js'
import { getThoughts } from '../../util.js'

it('edit thought', async () => {
  const thought = document.wrapper.find('div.editable')
  expect(thought.text()).toBe('')

  await thought.simulate('change', { target: { value: 'c' } })
  jest.runAllTimers()

  // for some reason the text of the Editable element is not updated
  // await document.wrapper.update()
  // expect(document.wrapper.find('div.editable').text()).toBe('c')

  const children = getThoughts([ROOT_TOKEN])
  expect(children).toHaveLength(1)
  expect(children[0]).toMatchObject({ value: 'c', rank: 0 })
})

```

**Comment by raineorshine** [comment]
These changes are important to set `EDIT_THROTTLE = 0` during testing and fix the auto merge:

https://github.com/cybersemics/em/commit/766e1a601f08db2f60a3d0b4896a9179553a1cd8

File: src/components/__tests__/ThoughtDelete.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+  await keyboardResponder.simulate('keydown', { key: 'Backspace' })
+  await thought.simulate('change', { target: { value: '' } })
+  await act(async () => {
+    await thought.simulate('keydown', { key: 'Backspace' })
+  })
+  jest.runAllTimers()
+  await document.wrapper.update()
+  const emptythoughts = document.wrapper.find('div.transformContain div ul.children')
+  expect(emptythoughts.length).toBe(0)
+})
```
Comment (raineorshine) [inline src/components/__tests__/ThoughtDelete.js]: I am finding that these longer tests without any comments or spacing are a bit hard to read. Can you kindly add line breaks and comments in between major actions, such as "new thought", "edit thought", and "delete thought"?

File: src/setupTests.js [hunk 20]
```diff
@@ -1,4 +1,78 @@
...
+    addRange: () => {},
+  })
+
+  window.location = {
+    pathname: '',
+  }
+
+  await act(async () => {
+    jest.useFakeTimers()
+    const originalError = console.error
```
Comment (raineorshine) [inline src/setupTests.js]: Hey, I'm not really comfortable with disabling `console.error` to hide the error. It seems like it's just a workaround. Let's try to get to the source of the problem. If it is impossible to remove, it deserves a clear comment explaining why.

Comment (mekanix) [reply src/setupTests.js]: I think https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOMLegacy.js#L81 is where we get that error from. As I don't see any way to go around it except creating new `console.error` which filters what it outputs by content (we can output only if first argument to `console.error` is not the one from the link).

File: src/components/__tests__/SubThoughtCreate.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+  expect(thought.text()).toBe('')
+  const keyboardResponder = document.wrapper.find('#keyboard')
+  await thought.simulate('change', { target: { value: 'c' } })
+  await keyboardResponder.simulate('keydown', { key: 'Enter', ctrlKey: true })
+  await thought.update()
+  jest.runAllTimers()
+  const children = getThoughts([ROOT_TOKEN])
+  expect(children).toHaveLength(1)
+  expect(children[0]).toMatchObject({ value: 'c', rank: 0 })
+  expect(getThoughtsRanked(['c'])[0].value).toBe('')
```
[the remark this answers is not in the document]
Comment (mekanix) [reply src/components/__tests__/SubThoughtCreate.js]: Tried it that way and tests are failing.

File: src/setupTests.js (hunk 20 above)
Comment (mekanix) [reply src/setupTests.js]: New `console.error` mock is more precise in what it silences. Good enough?

**Comment by raineorshine** [comment]
I was going to try [this solution](https://github.com/b52/electron-es6-react/issues/7#issue-192425844), but this branch is crashing as-is when I run `npm run test`. Is it working for you?

```js
  ● Console

    console.warn src/reducers/settings.js:25
      Missing oldThoughtRanked in Settings update: Tutorial Off


 RUNS  src/components/__tests__/SubThoughtCreate.js
/Users/raine/projects/em/node_modules/react-scripts/scripts/test.js:20
  throw err;
  ^

TypeError: Cannot read property '_location' of null
    at Window.get location [as location] (/Users/raine/projects/em/node_modules/jsdom/lib/jsdom/browser/Window.js:148:79)
    at Object.<anonymous>.exports.loadLocalState (/Users/raine/projects/em/src/util/loadLocalState.js:56:32)
npm ERR! Test failed.  See above for more details.
```

**Comment by mekanix** [comment]
I do see it in the log, but it doesn't break. This is the whole log on my machine http://dpaste.com/27JD3PE

**Comment by raineorshine** [comment]
Great! The `unmount` fixed it. All tests passing for me.

File: src/setupTests.js (hunk 11 above)
Comment (raineorshine) [reply src/setupTests.js]: Still waiting on this

File: src/setupTests.js (hunk 20 above)
Comment (raineorshine) [reply src/setupTests.js]: Mounting to a child of `document.body` will fix the error without the workaround:

```
    // mount to child to avoid warning about rendering directly to document.body
    const root = document.body.appendChild(document.createElement('div'))

    const wrapper = await mount(
      <div
        id="keyboard"
        onKeyDown={keyDown}
        tabIndex="0"
      >
        {app}
      </div>,
      { attachTo: root },
    )
```

File: src/components/__tests__/SubThoughtEdit.js [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  // create subthought
+  await keyboardResponder.simulate('keydown', { key: 'Enter', ctrlKey: true })
+  jest.runAllTimers()
+  await thought.update()
+  expect(getThoughtsRanked(['c'])[0].value).toBe('')
+
+  // find subthought
+  await document.wrapper.update()
+  const subthought = document.wrapper.find(
+    'ul.distance-from-cursor-0 div.editable',
```
[the remark this answers is not in the document]
Comment (mekanix) [reply src/components/__tests__/SubThoughtEdit.js]: It doesn't work with ul.children, I tried

File: src/components/__tests__/SubThoughtCreate.js [hunk 23]
```diff
@@ -0,0 +1,29 @@
+// import { getThoughtsRanked } from '../../util.js'
+import { ROOT_TOKEN } from '../../constants.js'
+import { getThoughts, getThoughtsRanked } from '../../util.js'
+
+it('create subthought', async () => {
+  try {
```
[the remark this answers is not in the document]
Comment (mekanix) [reply src/components/__tests__/SubThoughtCreate.js]: Yes :o(

**Comment by mekanix** [comment]
Github doesn't let me reply on last two comments.

`Still waiting on this`:
Isn't it more logical to have something in setup that will happen every single time? If not, I'll copy/paste it into tests, but it makes no sense to me.

`document.body` is not used any more

**Comment by raineorshine** [comment]
> Isn't it more logical to have something in setup that will happen every single time? If not, I'll copy/paste it into tests, but it makes no sense to me.

That's a fair argument I suppose. To me it seems clear that creating a thought is a testable action, not part of setup. It is also more future-proof, as there will likely be tests that do not begin by creating a new thought.

File: src/components/__tests__/SubThoughtEdit.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (raineorshine) [reply src/components/__tests__/SubThoughtEdit.js]: Let's use `.children .children div.editable`. I think the `children` class will be more resilient to change than `distance-from-cursor`.
